core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -310,6 +312,14 @@ public Node start() {
         // Start the transport service now so the publish address will be added to the local disco node in ClusterService
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.start();
+        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
+                .buildLocalNode(transportService.boundAddress().publishAddress());
+        clusterService.setLocalNode(localNode);
```

Comment:
instead of having this setter I think we should do this in `ClusterService` and `TransportService`

```Java

@Override
public void start() {
  throw new UnsupportedOpertationException("use start(DiscoveryNode) instead");
}

public void start(DiscoveryNode node) {
  this.setLocalNode(node); // private call
  super.start();
}
```

WDYT